File: shop.py
"""Shop: weapons, ammo, armor, health — paid from the persistent wallet."""
from __future__ import annotations

from save_manager import SaveManager
from weapon import WEAPON_DATA, WEAPON_ORDER

HEALTH_REFILL_PRICE = 150
MAX_HP_PRICE = 300          # +20 max HP (permanent this run)
ARMOR_PRICE = 500           # +10 armor
AMMO_PACK_PRICE = 150


class Shop:
    """Immediate-mode catalog; entries rendered by menu.py."""

    def get_entries(self, game) -> list[dict]:
        """Return purchasable rows for the current game state."""
        entries: list[dict] = []
        owned = set(game.player.weapons.weapons)
        for wid in WEAPON_ORDER:
            data = WEAPON_DATA[wid]
            if wid in owned:
                entries.append({"key": f"weapon:{wid}",
                                "label": data["name"],
                                "detail": f"DMG {data['damage']} x{data['pellets']}"
                                          f"  MAG {data['magazine']}",
                                "price": 0, "owned": True})
            else:
                entries.append({"key": f"weapon:{wid}",
                                "label": data["name"],
                                "detail": f"DMG {data['damage']} x{data['pellets']}"
                                          f"  MAG {data['magazine']}",
                                "price": data["price"], "owned": False})
        p = game.player
        entries.append({"key": "ammo_pack", "label": "AMMO PACK",
                        "detail": f"+{int(p.weapons.current.magazine_size * 3)}"
                                  f" reserve ({p.weapons.current.name})",
                        "price": AMMO_PACK_PRICE, "owned": False})
        entries.append({"key": "health", "label": "FULL HEAL",
                        "detail": f"HP {int(p.hp)}/{int(p.max_hp)}",
                        "price": HEALTH_REFILL_PRICE, "owned": False})
        entries.append({"key": "armor", "label": "ARMOR +10",
                        "detail": f"Armor {int(p.armor)}/100",
                        "price": ARMOR_PRICE, "owned": False})
        entries.append({"key": "max_hp", "label": "MAX HP +20",
                        "detail": f"Max HP {int(p.max_hp)}",
                        "price": MAX_HP_PRICE, "owned": False})
        return entries

    def buy(self, key: str, game) -> bool:
        """Attempt purchase; returns True on success.

        Persistent upgrades (max_hp, armor) are stored on the save profile
        and re-applied to every fresh run.
        """
        p = game.player
        save: SaveManager = game.save
        if key.startswith("weapon:"):
            wid = key.split(":", 1)[1]
            price = WEAPON_DATA[wid]["price"]
            if wid in p.weapons.weapons or p.coins < price:
                return False
            p.coins -= price
            p.weapons.give(wid)
            p.weapons.current_id = wid
            save.data.setdefault("unlocked_weapons", []).append(wid)
        else:
            prices = {"ammo_pack": AMMO_PACK_PRICE, "health": HEALTH_REFILL_PRICE,
                      "armor": ARMOR_PRICE, "max_hp": MAX_HP_PRICE}
            price = prices[key]
            if p.coins < price:
                return False
            p.coins -= price
            if key == "ammo_pack":
                w = p.weapons.current
                w.add_reserve(int(w.magazine_size * 3))
            elif key == "health":
                p.heal(p.max_hp)
            elif key == "armor":
                p.add_armor(10)
                save.data.setdefault("permanent_armor", 0)
                save.data["permanent_armor"] = min(
                    100, save.data["permanent_armor"] + 10)
                p.armor = max(p.armor, save.data["permanent_armor"])
            elif key == "max_hp":
                p.max_hp += 20
                p.heal(20)
                save.data.setdefault("permanent_max_hp", 0)
                save.data["permanent_max_hp"] += 20
        game.audio.play("buy")
        save.coins = p.coins
        save.save()
        game.toast("PURCHASED!")
        return True

File: web/src/game/shop.ts
// src/game/shop.ts
// Shop: weapon unlocks, ammo, health, armor, max HP. Mirrors shop.py.

import { WEAPON_ORDER } from "./weapon";
import { BOMB_PACK_AMOUNT, BOMB_PACK_PRICE } from "./grenade";
import type { WeaponData } from "./data";
import type { IGame } from "./types";

export const HEALTH_REFILL_PRICE = 150;
export const MAX_HP_PRICE = 300;
export const ARMOR_PRICE = 500;
export const AMMO_PACK_PRICE = 150;
export { BOMB_PACK_PRICE };
import { DRONE_PRICE, MAX_UFO_OWNED, firstUnownedUFO, ufoDef } from "./ufo";
export { DRONE_PRICE, MAX_UFO_OWNED, firstUnownedUFO, ufoDef };

export class Shop {
  data: Record<string, WeaponData>;

  constructor(data: Record<string, WeaponData>) {
    this.data = data;
  }

  /**
   * Buy one UFO. Enforces every rule server-side-of-the-client: unknown id,
   * already owned, fleet full (>= MAX_UFO_OWNED), and insufficient coins are
   * all rejected BEFORE any money moves — repeated/double clicks can never
   * double-charge or exceed the 4-UFO limit. Ownership is written to the
   * persistent profile (localStorage) and mirrored into the run save by
   * performSaveGame.
   */
  buyUFO(
    ufoId: string,
    game: IGame,
  ): "ok" | "owned" | "limit" | "funds" | "invalid" {
    const def = ufoDef(ufoId);
    if (!def) return "invalid";
    const p = game.player!;
    const save = game.save;
    if (p.ownedUFOs.includes(ufoId)) return "owned";
    if (p.ownedUFOs.length >= MAX_UFO_OWNED) return "limit";
    if (p.coins < def.price) return "funds";
    p.coins -= def.price;
    // First purchase auto-equips; later ones stay on the current active UFO.
    p.setUFOs([...p.ownedUFOs, ufoId], p.activeUFO || ufoId);
    save.data.owned_ufos = [...p.ownedUFOs];
    save.data.active_ufo = p.activeUFO;
    save.data.has_drone = p.hasDrone;
    save.coins = p.coins;
    save.save();
    game.audio.playSFX("ui.purchase", p.pos);
    return "ok";
  }

  /** Equip an owned UFO as the ACTIVE saucer (no cost). */
  equipUFO(ufoId: string, game: IGame): boolean {
    const p = game.player!;
    if (!p.ownedUFOs.includes(ufoId)) return false;
    p.setUFOs(p.ownedUFOs, ufoId);
    const save = game.save;
    save.data.owned_ufos = [...p.ownedUFOs];
    save.data.active_ufo = ufoId;
    save.data.has_drone = true;
    save.save();
    game.audio.playSFX("ui.purchase", p.pos);
    return true;
  }

  buy(key: string, game: IGame): boolean {
    const p = game.player!;
    const save = game.save;
    if (key === "drone") {
      // Legacy single-drone entry → first UFO purchase.
      return this.buyUFO("drone", game) === "ok";
    }
    if (key === "ufo_fleet") {
      // Legacy list shop's single UFO FLEET row → buy the next unowned UFO.
      const next = firstUnownedUFO(p.ownedUFOs);
      if (!next) return false;
      return this.buyUFO(next.id, game) === "ok";
    }
    if (key.startsWith("weapon:")) {
      const wid = key.slice("weapon:".length);
      const price = this.data[wid]?.price ?? 0;
      const owned = (p.weapons as unknown as { weapons: Record<string, unknown> }).weapons;
      if (owned[wid] || p.coins < price) return false;
      p.coins -= price;
      (p.weapons as unknown as { give: (id: string) => boolean }).give(wid);
      (p.weapons as unknown as { currentId: string }).currentId = wid;
      const list = save.data.unlocked_weapons;
      if (!list.includes(wid)) list.push(wid);
    } else {
      if (key === "drone" && p.hasDrone) return false;
      if (key === "bomb_pack" && p.bombs >= p.maxBombs) return false;
      const price = priceFor(key);
      if (p.coins < price) return false;
      p.coins -= price;
      if (key === "ammo_pack") {
        const w = p.weapons.current;
        w.addReserve(w.magazineSize * 3);
      } else if (key === "bomb_pack") {
        p.addBombs(BOMB_PACK_AMOUNT);
      } else if (key === "health") {
        p.heal(p.maxHp);
      } else if (key === "armor") {
        p.addArmor(10);
      } else if (key === "max_hp") {
        p.maxHp += 20;
        p.heal(20);
      } else {
        return false;
      }
    }
    game.audio.playSFX("ui.purchase", game.player!.pos);
    save.coins = p.coins;
    save.save();
    game.toast("PURCHASED!");
    return true;
  }
}

function priceFor(key: string): number {
  switch (key) {
    case "ammo_pack":
      return AMMO_PACK_PRICE;
    case "bomb_pack":
      return BOMB_PACK_PRICE;
    case "health":
      return HEALTH_REFILL_PRICE;
    case "armor":
      return ARMOR_PRICE;
    case "max_hp":
      return MAX_HP_PRICE;
    case "drone":
      return DRONE_PRICE;
    default:
      return Infinity;
  }
}

export function listWeapons() {
  return WEAPON_ORDER;
}
